lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1031 @@
+DEF_ALL_WRITER_NODE(Max)
+DEF_ALL_WRITER_NODE(Min)
+DEF_ALL_WRITER_NODE(Log)
+DEF_ALL_WRITER_NODE(Exp)
+DEF_ALL_WRITER_NODE(Relu)
+DEF_ALL_WRITER_NODE(Tanh)
+DEF_ALL_WRITER_NODE(Tile)
+DEF_ALL_WRITER_NODE(IsNaN)
+DEF_ALL_WRITER_NODE(Sigmoid)
+DEF_ALL_WRITER_NODE(LengthsSum)
+DEF_ALL_WRITER_NODE(BatchOneHot)
+DEF_ALL_WRITER_NODE(SparseToDense)
+DEF_ALL_WRITER_NODE(LengthsToRanges)
+DEF_ALL_WRITER_NODE(SparseLengthsSum)
+DEF_ALL_WRITER_NODE(SparseLengthsWeightedSum)
+DEF_ALL_WRITER_NODE(FusedRowwiseQuantizedSparseLengthsSum)
+
+// Glow nodes with default exporting algorithm.
+DEF_ALL_WRITER_NODE(CmpEQ)
+DEF_ALL_WRITER_NODE(CmpLTE)
+DEF_ALL_WRITER_NODE(Select)
+DEF_ALL_WRITER_NODE(Quantize)
+DEF_ALL_WRITER_NODE(ConvertTo)
+DEF_ALL_WRITER_NODE(BatchedAdd)
+DEF_ALL_WRITER_NODE(Dequantize)
+DEF_ALL_WRITER_NODE(Regression)
+DEF_ALL_WRITER_NODE(ScatterAssign)
+DEF_ALL_WRITER_NODE(IntLookupTable)
+DEF_ALL_WRITER_NODE(LengthsRangeFill)
+DEF_ALL_WRITER_NODE(RescaleQuantized)
+DEF_ALL_WRITER_NODE(RowwiseQuantizedFullyConnected)
+DEF_ALL_WRITER_NODE(RowwiseQuantizedSparseLengthsWeightedSum)
+DEF_ALL_WRITER_NODE(FusedRowwiseQuantizedSparseLengthsWeightedSum)
+DEF_ALL_WRITER_NODE(FullyConnected)
+
+// Unsupported for export Glow nodes.
+#define DEF_UNSUPPORTED_STORAGE(NAME)                                          \
+  llvm::Error ONNXModelWriter::write##NAME(const NAME *node, GraphType &,      \
+                                           ReporterType &) {                   \
+    return writeUnexpectedKind(node);                                          \
+  }
+
+// Helper nodes.
+DEF_UNSUPPORTED_STORAGE(Placeholder)
+DEF_UNSUPPORTED_STORAGE(Constant)
+DEF_UNSUPPORTED_STORAGE(Storage)
+
+// Unsupported for export Glow nodes.
+#define DEF_UNSUPPORTED_NODE(NAME)                                             \
+  llvm::Error ONNXModelWriter::write##NAME(const NAME##Node *node,             \
+                                           GraphType &, ReporterType &) {      \
+    return writeUnexpectedKind(node);                                          \
+  }
+
+DEF_UNSUPPORTED_NODE(SGD)
+// Artificial node.
+DEF_UNSUPPORTED_NODE(Save)
+// Gradient nodes.
+DEF_UNSUPPORTED_NODE(AddGrad)
+DEF_UNSUPPORTED_NODE(DivGrad)
+DEF_UNSUPPORTED_NODE(MulGrad)
+DEF_UNSUPPORTED_NODE(SubGrad)
+DEF_UNSUPPORTED_NODE(ReluGrad)
+DEF_UNSUPPORTED_NODE(TanhGrad)
+DEF_UNSUPPORTED_NODE(AvgPoolGrad)
+DEF_UNSUPPORTED_NODE(MaxPoolGrad)
+DEF_UNSUPPORTED_NODE(SigmoidGrad)
+DEF_UNSUPPORTED_NODE(SoftMaxGrad)
+DEF_UNSUPPORTED_NODE(RegressionGrad)
+DEF_UNSUPPORTED_NODE(ConvolutionGrad)
+DEF_UNSUPPORTED_NODE(CrossEntropyLoss)
+DEF_UNSUPPORTED_NODE(Convolution3DGrad)
+DEF_UNSUPPORTED_NODE(FullyConnectedGrad)
+DEF_UNSUPPORTED_NODE(CrossEntropyLossGrad)
+DEF_UNSUPPORTED_NODE(BatchNormalizationGrad)
+DEF_UNSUPPORTED_NODE(SparseLengthsWeightedSumGrad)
+DEF_UNSUPPORTED_NODE(SigmoidCrossEntropyWithLogits)
+DEF_UNSUPPORTED_NODE(LocalResponseNormalizationGrad)
+
+#if GLOW_WITH_OPENCL == 1
```

Comment:
I added a new OpenCL-only node type yesterday called `OCLBatchedReduceAdd` so you'll have to rebase and add another small function below to write that as well.

Reply:
Yeah, have seen it, working on it.